frappe/erpnext
# perf: BOM Update Tool

<details><summary> First Draft </summary>

##### `get_boms_in_bottom_up_order`
- Create child-parent map once and fetch value from child key to get parents
- Get parents recursively for a leaf node (get all ancestors)
- Approx. 44 secs for 4lakh 70k BOMS (used to go on for over 5-10 mins earlier, maybe more idk)

##### `update_cost`
- Remove double RM cost calculation
- Update child items in db as well (optionally)
- Don't regenerate exploded items for each BOM while updating cost (only in update cost job/bom function)🤦🏻‍♀️
- Approx 5 minutes for 20k BOMs. [WIP]
</details>

> There were correctness issues that could all be solved with a newer AlGoRiThM, that took preference. So first correctness >> then perf

### New BOM Cost Updation Flow
> **_Dependants_** = Immediate Parents, **_Dependencies_** = Children (parent-child is not accurate really)

Consider the following BOMs and their dependencies:

![BOM-structure drawio](https://user-images.githubusercontent.com/25857446/170047042-e85b90e1-5bdf-4d5b-926e-5bde51542e46.png)
- While updating all BOMs, we update the **Raw Material BOMs** or leaf nodes first and work our way up
- An important thing here: **We must only update a BOM after all its _Dependencies_ are resolved/updated**
- That means: α (alpha) cannot be updated until A,B,1,2,3,4 are updated
- For this we update BOMs **one level at a time**, and **within a level we process BOMs parallelly as they are independent.**
- BOMs **within each level** are split into batches of 20k (takes 5 mins) and queued for updation
- Once a level is done a Cron Job will prep and start processing the next level

#### Working:
- Given the above example, on submitting the BOM Update Log, **Raw Material BOMs** will be processed first (batch size = 20k). The status will be **In Progress**

  <img src="https://user-images.githubusercontent.com/25857446/170224003-1d87b8cd-9264-4b02-8da0-30e37255465b.png" width="60%">
- These Jobs will keep updating their row in the table in the Log and keep checking if all other jobs of this level are done (via `bom_batches` table)
- The **Current Level** is also maintained in the Progress Section, which will increment as we progress up the levels
- **Once a level is complete** and the cron (every 5 mins) determines a level is complete (all rows of current level have status **Pending** -> **Completed**), it will prepare next level parent boms (dependants) and kickstart the next level

   <img width="70%" alt="Screenshot 2022-05-24 at 2 17 49 PM" src="https://user-images.githubusercontent.com/25857446/172573240-12d10bb3-c910-4dd0-bfb4-26facf74d714.png">
- **How do we determine "Parent BOMs" ?** : By checking immediate parents with no **_Dependencies_**. Here **1** has immediate parents **A** and **α (alpha)**. **A** can be considered since it has no unresolved/non-updated **_Dependencies_**, but **α (alpha)** has  A and B as unresolved dependencies. **So we proceed with A as parent of 1 in the next level**.
- A Cron Job will run every 5 minutes and check for **In Progress** jobs and start processing the next level **if all jobs of current level are completed**
- The Log will be set as **Complete** when there are no more **_Dependants_/ Parent BOMs** to resolve/update

-----------------------------
 > #### Important change/fix to test
> - Rate updation as per **Raw Material Cost as per** used to not work, which now does
> - Rates are reset rightfully on save and fetched as per Valuation/Last Purchase/Price List
----------------------------
**To Test:**
- Create BOM with 3 levels (RM BOM, Subassembly BOM that uses RM BOM, FG BOM that uses Subassemble BOM)
- Set RM cost as per **Valuation Rate** > Change valuation rate of Item in RM BOM > Update all costs via tool > check
- Set RM cost as per **Valuation Rate** with different BOM currency (USD)> Change valuation rate of Item in RM BOM > Update all costs via tool > check
- Set RM cost as per **Price List** > Change Price List rate of Item in RM BOM > Update all costs via tool > check
- Set RM cost as per **Price List** in different Price List currency > Change Price List rate of Item in RM BOM > Update all costs via tool > check

**Todo:**

- [x] Clearing BOM Update Logs periodically (or atleast the table in it)
